crates/ra_ide/src/completion/complete_trait_impl.rs

```diff
@@ -0,0 +1,356 @@
+use crate::completion::{
+    CompletionContext, CompletionItem, CompletionItemKind, CompletionKind, Completions,
+};
+
+use ra_syntax::{ast::{self, edit}, AstNode, SyntaxKind, TextRange};
+use hir::{self, Docs, HasSource};
+
+use ra_assists::utils::get_missing_impl_items;
+
+/// Analyzes the specified `CompletionContext` and provides magic completions
+/// if the context falls within a `impl Trait for` block.
+/// 
+/// # Completion Activation
+/// The completion will activate when a user begins to type a function 
+/// definition, an associated type, or an associated constant.
+/// 
+/// ### Functions
+/// ```ignore
+/// trait SomeTrait {
+///     fn foo(&self);
+/// }
+/// 
+/// impl SomeTrait for () {
+///     fn <|>
+/// }
+/// ```
+/// 
+/// ### Associated Types
+/// ```ignore
+/// trait SomeTrait {
+///     type SomeType;
+/// }
+/// 
+/// impl SomeTrait for () {
+///     type <|>
+/// }
+/// ```
+/// 
+/// ### Associated Constants
+/// ```ignore
+/// trait SomeTrait {
+///     const SOME_CONST: u16;
+/// }
+/// 
+/// impl SomeTrait for () {
+///     const <|>
+/// }
+/// ```
+pub(crate) fn complete_trait_impl(acc: &mut Completions, ctx: &CompletionContext) {
+
+    // it is possible to have a parent `fn` and `impl` block. Ignore completion 
+    // attempts from within a `fn` block.
+    if ctx.function_syntax.is_some() {
+        return;
+    }
```

Comment:
I think we might want to flip polarity of the check here. As written, it's not obvious why completion won't trigger in some cases where it shouldn't. Like for exmple in `const T: u32 = { foo<cursor here>() };`.

I think the logic should be: check if the current token is one of `fn`, `const`, `type`, `ident` and that it's parent is an impl block. 

Ie, by default we *don't* provide these completions

Reply:
Ah, makes sense. Would there any need for the `CompletionContext` to provide the `impl_block` as this behavior seems pretty specific to this completion type?

Reply:
As long as I interpreted your suggestion correctly, I think it is done. It now essentially uses a `FN_DEF`/`TYPE_ALIAS_DEF`/`CONST_DEF` as a trigger into the completion. As a side effect it seems to prevent the magic completion from showing up until the user starts writing an `IDENT`, even if cursor is within an `FN_DEF`/`TYPE_ALIAS_DEF`. So it seems to be closer a VS Code issue maybe. That being said, I actually think I like it this way anyways.

I've update the gif in the summary that demonstrates the new behavior.